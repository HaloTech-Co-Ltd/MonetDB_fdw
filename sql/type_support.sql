select monetdb_execute('foreign_server', $$CREATE TABLE Numeric_Types(
	a TINYINT,
	b SMALLINT,
	c INTEGER,
	d BIGINT,
	-- e HUGEINT, 
	f DECIMAL,
	g NUMERIC(38, 3),
	h REAL,
	i DOUBLE PRECISION,
	j FLOAT 
)$$);

IMPORT FOREIGN SCHEMA "test_u" limit to (Numeric_Types) from server foreign_server into public; 
\d+ Numeric_Types

-- test TINYINT
INSERT INTO Numeric_Types(a) VALUES(-127 - 1);
INSERT INTO Numeric_Types(a) VALUES(127);
INSERT INTO Numeric_Types(a) VALUES(127 + 1);
SELECT * FROM Numeric_Types;

TRUNCATE Numeric_Types;
-- test SMALLINT
INSERT INTO Numeric_Types(b) VALUES(-32767 - 1);
INSERT INTO Numeric_Types(b) VALUES(32767);
INSERT INTO Numeric_Types(b) VALUES(32767 + 1);
SELECT * FROM Numeric_Types;

TRUNCATE Numeric_Types;
-- test INTEGER
INSERT INTO Numeric_Types(c) VALUES(-2147483647 - 1);
INSERT INTO Numeric_Types(c) VALUES(2147483647);
INSERT INTO Numeric_Types(c) VALUES(2147483647 + 1);
SELECT * FROM Numeric_Types;

TRUNCATE Numeric_Types;
-- test BIGINT
INSERT INTO Numeric_Types(d) VALUES(-9223372036854775807 - 1);
INSERT INTO Numeric_Types(d) VALUES(9223372036854775807);
INSERT INTO Numeric_Types(d) VALUES(9223372036854775807 + 1);
SELECT * FROM Numeric_Types;

select monetdb_execute('foreign_server', $$CREATE TABLE bool_Types(
	a BOOLEAN
)$$);

IMPORT FOREIGN SCHEMA "test_u" limit to (bool_Types) from server foreign_server into public; 
\d+ bool_Types

-- test BOOLEAN
INSERT INTO bool_Types VALUES(true);
INSERT INTO bool_Types VALUES('true');
INSERT INTO bool_Types VALUES('1');
INSERT INTO bool_Types VALUES('t');
INSERT INTO bool_Types VALUES(false);
INSERT INTO bool_Types VALUES('false');
INSERT INTO bool_Types VALUES('0');
INSERT INTO bool_Types VALUES('f');
SELECT * FROM bool_Types;

DROP FOREIGN TABLE bool_Types;
DROP FOREIGN TABLE Numeric_Types;
SELECT monetdb_execute('foreign_server', $$DROP TABLE bool_Types$$);
SELECT monetdb_execute('foreign_server', $$DROP TABLE Numeric_Types$$);
SELECT monetdb_execute('foreign_server', $$ALTER USER test_u SET SCHEMA sys$$);
SELECT monetdb_execute('foreign_server', $$DROP SCHEMA test_u$$);
SELECT monetdb_execute('foreign_server', $$DROP USER test_u$$);