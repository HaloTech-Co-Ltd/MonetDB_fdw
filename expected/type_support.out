select monetdb_execute('foreign_server', $$CREATE TABLE Numeric_Types(
	a TINYINT,
	b SMALLINT,
	c INTEGER,
	d BIGINT,
	-- e HUGEINT, 
	f DECIMAL,
	g NUMERIC(38, 3),
	h REAL,
	i DOUBLE PRECISION,
	j FLOAT 
)$$);
 monetdb_execute 
-----------------
 
(1 row)

IMPORT FOREIGN SCHEMA "test_u" limit to (Numeric_Types) from server foreign_server into public; 
\d+ Numeric_Types
                                      Foreign table "public.numeric_types"
 Column |       Type       | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+------------------+-----------+----------+---------+-------------+---------+--------------+-------------
 a      | tinyint          |           |          |         |             | plain   |              | 
 b      | smallint         |           |          |         |             | plain   |              | 
 c      | integer          |           |          |         |             | plain   |              | 
 d      | bigint           |           |          |         |             | plain   |              | 
 f      | numeric(18,3)    |           |          |         |             | main    |              | 
 g      | numeric(38,3)    |           |          |         |             | main    |              | 
 h      | real             |           |          |         |             | plain   |              | 
 i      | double precision |           |          |         |             | plain   |              | 
 j      | double precision |           |          |         |             | plain   |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'numeric_types')

-- test TINYINT
INSERT INTO Numeric_Types(a) VALUES(-127 - 1);
ERROR:  value for domain tinyint violates check constraint "tinyint_check"
INSERT INTO Numeric_Types(a) VALUES(127);
INSERT INTO Numeric_Types(a) VALUES(127 + 1);
ERROR:  value for domain tinyint violates check constraint "tinyint_check"
SELECT * FROM Numeric_Types;
  a  | b | c | d | f | g | h | i | j 
-----+---+---+---+---+---+---+---+---
 127 |   |   |   |   |   |   |   |  
(1 row)

TRUNCATE Numeric_Types;
-- test SMALLINT
INSERT INTO Numeric_Types(b) VALUES(-32767 - 1);
ERROR:  [MonetDB RESULT ERROR] conversion of string '-32768' to type sht failed.

INSERT INTO Numeric_Types(b) VALUES(32767);
INSERT INTO Numeric_Types(b) VALUES(32767 + 1);
ERROR:  smallint out of range
SELECT * FROM Numeric_Types;
 a |   b   | c | d | f | g | h | i | j 
---+-------+---+---+---+---+---+---+---
   | 32767 |   |   |   |   |   |   |  
(1 row)

TRUNCATE Numeric_Types;
-- test INTEGER
INSERT INTO Numeric_Types(c) VALUES(-2147483647 - 1);
ERROR:  [MonetDB RESULT ERROR] conversion of string '-2147483648' to type int failed.

INSERT INTO Numeric_Types(c) VALUES(2147483647);
INSERT INTO Numeric_Types(c) VALUES(2147483647 + 1);
ERROR:  integer out of range
SELECT * FROM Numeric_Types;
 a | b |     c      | d | f | g | h | i | j 
---+---+------------+---+---+---+---+---+---
   |   | 2147483647 |   |   |   |   |   |  
(1 row)

TRUNCATE Numeric_Types;
-- test BIGINT
INSERT INTO Numeric_Types(d) VALUES(-9223372036854775807 - 1);
ERROR:  [MonetDB RESULT ERROR] conversion of string '-9223372036854775808' to type lng failed.

INSERT INTO Numeric_Types(d) VALUES(9223372036854775807);
INSERT INTO Numeric_Types(d) VALUES(9223372036854775807 + 1);
ERROR:  bigint out of range
SELECT * FROM Numeric_Types;
 a | b | c |          d          | f | g | h | i | j 
---+---+---+---------------------+---+---+---+---+---
   |   |   | 9223372036854775807 |   |   |   |   |  
(1 row)

select monetdb_execute('foreign_server', $$CREATE TABLE bool_Types(
	a BOOLEAN
)$$);
 monetdb_execute 
-----------------
 
(1 row)

IMPORT FOREIGN SCHEMA "test_u" limit to (bool_Types) from server foreign_server into public; 
\d+ bool_Types
                                   Foreign table "public.bool_types"
 Column |  Type   | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+---------+-----------+----------+---------+-------------+---------+--------------+-------------
 a      | boolean |           |          |         |             | plain   |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'bool_types')

-- test BOOLEAN
INSERT INTO bool_Types VALUES(true);
INSERT INTO bool_Types VALUES('true');
INSERT INTO bool_Types VALUES('1');
INSERT INTO bool_Types VALUES('t');
INSERT INTO bool_Types VALUES(false);
INSERT INTO bool_Types VALUES('false');
INSERT INTO bool_Types VALUES('0');
INSERT INTO bool_Types VALUES('f');
SELECT * FROM bool_Types;
 a 
---
 t
 t
 t
 t
 f
 f
 f
 f
(8 rows)

DROP FOREIGN TABLE bool_Types;
DROP FOREIGN TABLE Numeric_Types;
SELECT monetdb_execute('foreign_server', $$DROP TABLE bool_Types$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$DROP TABLE Numeric_Types$$);
 monetdb_execute 
-----------------
 
(1 row)

select monetdb_execute('foreign_server', $$CREATE TABLE Character_Types(
	a CHAR,
	b CHARACTER(10),
	c VARCHAR(20),
	d VARCHAR,
	e CLOB,
	f TEXT,  -- The TEXT type of the external table imported using IMPORT is not the TEXT type of PostgrSQL
	g STRING,
	h CLOB(10),
	i TEXT(20),
	j STRING(30) 
)$$);
 monetdb_execute 
-----------------
 
(1 row)

IMPORT FOREIGN SCHEMA "test_u" limit to (Character_Types) from server foreign_server into public; 
\d+ Character_Types
                                        Foreign table "public.character_types"
 Column |         Type          | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------------------+-----------+----------+---------+-------------+----------+--------------+-------------
 a      | character(1)          |           |          |         |             | extended |              | 
 b      | character(10)         |           |          |         |             | extended |              | 
 c      | character varying(20) |           |          |         |             | extended |              | 
 d      | character varying     |           |          |         |             | extended |              | 
 e      | character varying     |           |          |         |             | extended |              | 
 f      | character varying     |           |          |         |             | extended |              | 
 g      | character varying     |           |          |         |             | extended |              | 
 h      | character varying(10) |           |          |         |             | extended |              | 
 i      | character varying(20) |           |          |         |             | extended |              | 
 j      | character varying(30) |           |          |         |             | extended |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'character_types')

-- In fact, CLOB, TEXT, and STRING with precision information are not supported
CREATE TABLE Character_Types2(
	e CLOB,
	f TEXT,
	g STRING
);
DROP TABLE Character_Types2;
-- 
CREATE TABLE Character_Types2(h CLOB(10));      -- error
ERROR:  type modifier is not allowed for type "clob"
LINE 1: CREATE TABLE Character_Types2(h CLOB(10));
                                        ^
CREATE TABLE Character_Types2(i TEXT(20));      -- error
ERROR:  type modifier is not allowed for type "text"
LINE 1: CREATE TABLE Character_Types2(i TEXT(20));
                                        ^
CREATE TABLE Character_Types2(j STRING(30));    -- error
ERROR:  type modifier is not allowed for type "string"
LINE 1: CREATE TABLE Character_Types2(j STRING(30));
                                        ^
DROP FOREIGN TABLE Character_Types;
-- Create an external table with the following PG types
CREATE FOREIGN TABLE Character_Types(
	a CHAR,
	b CHARACTER(10),
	c VARCHAR(20),
	d VARCHAR,
	e CLOB,
	f TEXT,
	g STRING,
	h VARCHAR(10),
	i VARCHAR(20),
	j VARCHAR(30) 
)
SERVER foreign_server
OPTIONS (schema_name 'test_u', table_name 'Character_Types');
\d+ Character_Types
                                        Foreign table "public.character_types"
 Column |         Type          | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+-----------------------+-----------+----------+---------+-------------+----------+--------------+-------------
 a      | character(1)          |           |          |         |             | extended |              | 
 b      | character(10)         |           |          |         |             | extended |              | 
 c      | character varying(20) |           |          |         |             | extended |              | 
 d      | character varying     |           |          |         |             | extended |              | 
 e      | clob                  |           |          |         |             | extended |              | 
 f      | text                  |           |          |         |             | extended |              | 
 g      | string                |           |          |         |             | extended |              | 
 h      | character varying(10) |           |          |         |             | extended |              | 
 i      | character varying(20) |           |          |         |             | extended |              | 
 j      | character varying(30) |           |          |         |             | extended |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'Character_Types')

DROP FOREIGN TABLE Character_Types;
SELECT monetdb_execute('foreign_server', $$DROP TABLE Character_Types$$);
 monetdb_execute 
-----------------
 
(1 row)

select monetdb_execute('foreign_server', $$CREATE TABLE Time_Types(
	a TIMESTAMP,
	b TIMESTAMP WITH TIME ZONE,
	c DATE,
	d TIME,
	e TIME WITH TIME ZONE
)$$);
 monetdb_execute 
-----------------
 
(1 row)

IMPORT FOREIGN SCHEMA "test_u" limit to (Time_Types) from server foreign_server into public; 
\d+ Time_Types
                                             Foreign table "public.time_types"
 Column |            Type             | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+-----------------------------+-----------+----------+---------+-------------+---------+--------------+-------------
 a      | timestamp without time zone |           |          |         |             | plain   |              | 
 b      | timestamp with time zone    |           |          |         |             | plain   |              | 
 c      | date                        |           |          |         |             | plain   |              | 
 d      | time without time zone      |           |          |         |             | plain   |              | 
 e      | time with time zone         |           |          |         |             | plain   |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'time_types')

INSERT INTO time_types VALUES('2014-04-24 17:12:12.415', '2014-04-24 17:12:12.415 -02:00', '2014-04-24', '17:12:12.415', '17:12:12.415 -02:00');
SELECT * FROM time_types;
              a               |                b                 |     c      |    d     |      e      
------------------------------+----------------------------------+------------+----------+-------------
 Thu Apr 24 17:12:12.415 2014 | Thu Apr 24 12:12:12.415 2014 PDT | 04-24-2014 | 17:12:12 | 03:12:12+08
(1 row)

DROP FOREIGN TABLE Time_Types;
SELECT monetdb_execute('foreign_server', $$DROP TABLE Time_Types$$);
 monetdb_execute 
-----------------
 
(1 row)

select monetdb_execute('foreign_server', $$CREATE TABLE json_example (c1 JSON, c2 JSON(512) NOT NULL)$$);
 monetdb_execute 
-----------------
 
(1 row)

IMPORT FOREIGN SCHEMA "test_u" limit to (json_example) from server foreign_server into public; 
\d+ json_example
                                 Foreign table "public.json_example"
 Column | Type | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | json |           |          |         |             | extended |              | 
 c2     | json |           | not null |         |             | extended |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'json_example')

INSERT INTO json_example values(
'{ "store": {
    "book": [
      { "category": "reference",
        "author": "Nigel Rees",
        "title": "Sayings of the Century",
        "price": 8.95
      },
      { "category": "fiction",
        "author": "Evelyn Waugh",
        "title": "Sword of Honour",
        "price": 12.99
      }
    ],
    "bicycle": {
      "color": "red",
      "price": 19.95
    }
  }
}',
'{ "store": {
    "book": [
      { "category": "reference",
        "author": "Nigel Rees",
        "title": "Sayings of the Century",
        "price": 8.95
      },
      { "category": "fiction",
        "author": "Evelyn Waugh",
        "title": "Sword of Honour",
        "price": 12.99
      }
    ],
    "bicycle": {
      "color": "red",
      "price": 19.95
    }
  }
}');
SELECT * FROM json_example;
                                                                                                                        c1                                                                                                                        |                                                                                                                        c2                                                                                                                        
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"store":{"book":[{"category":"reference","author":"Nigel Rees","title":"Sayings of the Century","price":8.95},{"category":"fiction","author":"Evelyn Waugh","title":"Sword of Honour","price":12.99}],"bicycle":{"color":"red","price":19.95}}} | {"store":{"book":[{"category":"reference","author":"Nigel Rees","title":"Sayings of the Century","price":8.95},{"category":"fiction","author":"Evelyn Waugh","title":"Sword of Honour","price":12.99}],"bicycle":{"color":"red","price":19.95}}}
(1 row)

SELECT jsonb_path_query(c1::jsonb, '$.store') FROM json_example;  -- use postgresql function
                                                                                                                      jsonb_path_query                                                                                                                       
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"book": [{"price": 8.95, "title": "Sayings of the Century", "author": "Nigel Rees", "category": "reference"}, {"price": 12.99, "title": "Sword of Honour", "author": "Evelyn Waugh", "category": "fiction"}], "bicycle": {"color": "red", "price": 19.95}}
(1 row)

CREATE TABLE json_example2(h JSON(512));      -- error
ERROR:  type modifier is not allowed for type "json"
LINE 1: CREATE TABLE json_example2(h JSON(512));
                                     ^
DROP FOREIGN TABLE json_example;
SELECT monetdb_execute('foreign_server', $$DROP TABLE json_example$$);
 monetdb_execute 
-----------------
 
(1 row)

select monetdb_execute('foreign_server', $$CREATE TABLE uuid_example(a UUID)$$);
 monetdb_execute 
-----------------
 
(1 row)

IMPORT FOREIGN SCHEMA "test_u" limit to (uuid_example) from server foreign_server into public; 
\d+ uuid_example
                                 Foreign table "public.uuid_example"
 Column | Type | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+------+-----------+----------+---------+-------------+---------+--------------+-------------
 a      | uuid |           |          |         |             | plain   |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'uuid_example')

INSERT INTO uuid_example VALUES('26d7a80b-7538-4682-a49a-9d0f9676b765');
SELECT * FROM uuid_example;
                  a                   
--------------------------------------
 26d7a80b-7538-4682-a49a-9d0f9676b765
(1 row)

TRUNCATE uuid_example;
INSERT INTO uuid_example VALUES(gen_random_uuid());
SELECT count(*) FROM uuid_example;
 count 
-------
     1
(1 row)

DROP FOREIGN TABLE uuid_example;
SELECT monetdb_execute('foreign_server', $$DROP TABLE uuid_example$$);
 monetdb_execute 
-----------------
 
(1 row)

select monetdb_execute('foreign_server', $$CREATE TABLE inet_example(a inet)$$);
 monetdb_execute 
-----------------
 
(1 row)

IMPORT FOREIGN SCHEMA "test_u" limit to (inet_example) from server foreign_server into public; 
\d+ inet_example
                                 Foreign table "public.inet_example"
 Column | Type | Collation | Nullable | Default | FDW options | Storage | Stats target | Description 
--------+------+-----------+----------+---------+-------------+---------+--------------+-------------
 a      | inet |           |          |         |             | main    |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'inet_example')

INSERT INTO inet_example VALUES('192.168.1.5/24');
SELECT * FROM inet_example;
       a        
----------------
 192.168.1.5/24
(1 row)

DROP FOREIGN TABLE inet_example;
SELECT monetdb_execute('foreign_server', $$DROP TABLE inet_example$$);
 monetdb_execute 
-----------------
 
(1 row)

select monetdb_execute('foreign_server', $$CREATE TABLE URL_example (c1 URL, c2 URL(512) NOT NULL)$$);
 monetdb_execute 
-----------------
 
(1 row)

IMPORT FOREIGN SCHEMA "test_u" limit to (URL_example) from server foreign_server into public; 
\d+ URL_example
                                  Foreign table "public.url_example"
 Column | Type | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+------+-----------+----------+---------+-------------+----------+--------------+-------------
 c1     | url  |           |          |         |             | extended |              | 
 c2     | url  |           | not null |         |             | extended |              | 
Server: foreign_server
FDW options: (schema_name 'test_u', table_name 'url_example')

INSERT INTO url_example VALUES('https://github.com/Z-Xiao-M/MonetDB_fdw', 'https://github.com/Z-Xiao-M/MonetDB_fdw');
SELECT * FROM URL_example;
                   c1                    |                   c2                    
-----------------------------------------+-----------------------------------------
 https://github.com/Z-Xiao-M/MonetDB_fdw | https://github.com/Z-Xiao-M/MonetDB_fdw
(1 row)

DROP FOREIGN TABLE URL_example;
SELECT monetdb_execute('foreign_server', $$DROP TABLE URL_example$$);
 monetdb_execute 
-----------------
 
(1 row)

CREATE TABLE URL_example(h URL(512));      -- error
ERROR:  type modifier is not allowed for type "url"
LINE 1: CREATE TABLE URL_example(h URL(512));
                                   ^
SELECT monetdb_execute('foreign_server', $$ALTER USER test_u SET SCHEMA sys$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$DROP SCHEMA test_u$$);
 monetdb_execute 
-----------------
 
(1 row)

SELECT monetdb_execute('foreign_server', $$DROP USER test_u$$);
 monetdb_execute 
-----------------
 
(1 row)

